libraries/standard/mqtt/src/iot_mqtt_helper.c

```diff
@@ -80,8 +83,96 @@
 
 /*-----------------------------------------------------------*/
 
+/* Username for metrics with AWS IoT. */
+#if ( AWS_IOT_MQTT_ENABLE_METRICS == 1 ) || ( DOXYGEN == 1 )
+    #ifndef AWS_IOT_METRICS_USERNAME
+
+/**
+ * @brief Specify C SDK and version.
+ */
+        #define AWS_IOT_METRICS_USERNAME           "?SDK=C&Version=4.0.0"
+
+/**
+ * @brief The length of #AWS_IOT_METRICS_USERNAME.
+ */
+        #define AWS_IOT_METRICS_USERNAME_LENGTH    ( ( uint16_t ) sizeof( AWS_IOT_METRICS_USERNAME ) - 1U )
+    #endif /* ifndef AWS_IOT_METRICS_USERNAME */
+#endif /* if AWS_IOT_MQTT_ENABLE_METRICS == 1 || DOXYGEN == 1 */
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Encode both connection and metrics username into a buffer,
+ * if they will fit.
+ *
+ * @param[in] pDestination Buffer to write username into.
+ * @param[in] pConnectInfo User-provided CONNECT information.
+ * @param[out] pEncodedUserName Whether the username was written into the buffer.
+ *
+ * @return Pointer to the end of encoded string, which will be identical to
+ * `pDestination` if nothing was encoded.
+ *
+ * @warning This function does not check the size of `pDestination`! Ensure that
+ * `pDestination` is large enough to hold `pConnectInfo->userNameLength` +
+ * #AWS_IOT_METRICS_USERNAME_LENGTH bytes to avoid a buffer overflow.
+ */
+
+static uint8_t * _encodeUserNameAndMetrics( uint8_t * pDestination,
+                                            const IotMqttConnectInfo_t * pConnectInfo,
+                                            bool * pEncodedUserName );
+
+/*-----------------------------------------------------------*/
+
+static uint8_t * _encodeUserNameAndMetrics( uint8_t * pDestination,
+                                            const IotMqttConnectInfo_t * pConnectInfo,
+                                            bool * pEncodedUserName )
+{
+    uint8_t * pBuffer = pDestination;
+
+    #if AWS_IOT_MQTT_ENABLE_METRICS == 1
```

Comment:
Do we test with both metrics enabled and disabled?

Reply:
It depends on the configuration. I think we test with enabled , but my local setup it was disabled. So didn't this problem.